### PR TITLE
DAOS-14315 vos: Pin objects for DTX commit & CPD RPC

### PR DESCRIPTION
**Pull request by NiuYawei**
Introduced two new VOS APIs vos_pin_objects() & vos_unpin_objects()
for pin or unpin objects. Changed DTX commit & CPD RPC handler code
to ensure objects pinned before starting local transaction.

Bug fix in vos_pmemobj_create(), the actual scm_size should be passed
to bio_mc_create().

Required-githooks: true

Signed-off-by: Niu Yawei <yawei.niu@intel.com>

### Before requesting gatekeeper:

* [ ] Two review approvals and any prior change requests have been resolved.
* [ ] Testing is complete and all tests passed or there is a reason documented in the PR why it should be force landed and forced-landing tag is set.
* [ ] `Features:` (or `Test-tag*`) commit pragma was used or there is a reason documented that there are no appropriate tags for this PR.
* [ ] Commit messages follows the guidelines outlined [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments).
* [ ] Any tests skipped by the ticket being addressed have been run and passed in the PR.

### Gatekeeper:

* [ ] You are the appropriate gatekeeper to be landing the patch.
* [ ] The PR has 2 reviews by people familiar with the code, including appropriate owners.
* [ ] Githooks were used. If not, request that user install them and check copyright dates.
* [ ] Checkpatch issues are resolved.  Pay particular attention to ones that will show up on future PRs.
* [ ] All builds have passed.  Check non-required builds for any new compiler warnings.
* [ ] Sufficient testing is done. Check feature pragmas and test tags and that tests skipped for the ticket are run and now pass with the changes.
* [ ] If applicable, the PR has addressed any potential version compatibility issues.
* [ ] Check the target branch.   If it is master branch, should the PR go to a feature branch?  If it is a release branch, does it have merge approval in the JIRA ticket.
* [ ] Extra checks if forced landing is requested
  * [ ] Review comments are sufficiently resolved, particularly by prior reviewers that requested changes.
  * [ ] No new NLT or valgrind warnings.  Check the classic view.
  * [ ] Quick-build or Quick-functional is not used.
* [ ] Fix the commit message upon landing. Check the standard [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments). Edit it to create a single commit. If necessary, ask submitter for a new summary.

### CODE REVIEW

File: src/object/srv_obj.c [hunk 1]
```diff
@@ -4391,6 +4391,44 @@ obj_cpd_reply(crt_rpc_t *rpc, int status, uint32_t map_version)
...
+
+	for (i = 0; i < count; i++) {
+		dcsr = &dcsrs[dcri[i].dcri_req_idx];
+		dcsr->dcsr_oid.id_shard = dcri[i].dcri_shard_id;
+
+		D_ASSERT(dcsr->dcsr_opc != DCSO_READ);
+		oids[i] = dcsr->dcsr_oid;
+	}
+
+	rc = vos_pin_objects(coh, oids, count, pin_hdl);
```
Comment (Nasf-Fan) [inline src/object/srv_obj.c:4424]: Pin the whole object shard? It seems too much overhead. Is it possible to only pin the items (key/rec) to be modified? How regular update RPC handler to handle such kind of pin?

File: src/vos/vos_dtx.c [hunk 2]
```diff
@@ -2215,14 +2215,139 @@ vos_dtx_post_handle(struct vos_container *cont,
...
+
+	for (i = count - 1; i >= 0; i--) {
+		rec_off = umem_off2offset(DAE_REC_INLINE(dae)[i]);
+		rc = bkts_add_rec(pool, bkts, rec_off);
+		if (rc)
+			goto out;
+	}
+
+	/* Stop adding the dae when current dae not located in the subset of @bkts_in */
+	if (local_bkts.vba_cnt != 0 && !vos_bkt_array_subset(bkts_in, &local_bkts))
```
Comment (Nasf-Fan) [inline src/vos/vos_dtx.c:2279]: It means if former `dae` and current `dae` do not share the same `bkt_array`, then stop adding, right? If yes, it seems some inefficient. Can we enhance that to check whether the two `bkt_array` shares something or not?

File: src/vos/vos_dtx.c [hunk 3]
```diff
@@ -2215,14 +2215,139 @@ vos_dtx_post_handle(struct vos_container *cont,
...
+	if (rc)
+		DL_ERROR(rc, "Failed to pin buckets.");
+out:
+	vos_bkt_array_fini(&bkts);
+	return rc;
+}
+
 int
 vos_dtx_commit(daos_handle_t coh, struct dtx_id dtis[], int count, bool rm_cos[])
 {
```
Comment (Nasf-Fan) [inline src/vos/vos_dtx.c:2344]: Not only vos_dtx_commit() but also vos_dtx_abort() and vos_dtx_set_flags() need to pin related items to be modified.

File: src/vos/vos_dtx.c [hunk 4]
```diff
@@ -2237,24 +2362,42 @@ vos_dtx_commit(daos_handle_t coh, struct dtx_id dtis[], int count, bool rm_cos[]
...
 		}
-		vos_dtx_post_handle(cont, daes, dces, count, false, rc != 0);
+		vos_dtx_post_handle(cont, &daes[idx], &dces[idx], pinned, false, rc != 0);
 	}
 
+	dtx_commit_unpin(cont, pin_hdl);
+
+	if (count > 0) {
+		idx += pinned;
+		goto pin_objects;
```
Comment (Nasf-Fan) [inline src/vos/vos_dtx.c:2394]: So one vos_dtx_commit() will be handled via `N` (<= count) local TX, right? It seems work, but may be inefficient if the DTX entries to be committed all have no shares. But of course, depends on the which overhead is more large: local TX or pin/unpin the items.

File: src/vos/vos_io.c [hunk 5]
```diff
@@ -2447,7 +2447,8 @@ vos_update_end(daos_handle_t ioh, uint32_t pm_ver, daos_key_t *dkey, int err,
 	tx_started = true;
 
 	/* Commit the CoS DTXs via the IO PMDK transaction. */
-	if (dtx_is_valid_handle(dth) && dth->dth_dti_cos_count > 0 && !dth->dth_cos_done) {
+	if (dtx_is_valid_handle(dth) && dth->dth_dti_cos_count > 0 && !dth->dth_cos_done &&
+	    !vos_pool_is_evictable(vos_cont2pool(ioc->ic_cont))) {
```
Comment (Nasf-Fan) [inline src/vos/vos_io.c:2451]: Disable DTX commit-on-share (CoS) if pool is evictable? Some potential effect for that:
1. Some unexpected DER_INPROGRESS failure on non-leader under the case of leader does successfully.
2. Decrease the efficiency of DTX commit.

File: src/vos/vos_obj_cache.c [hunk 6]
```diff
@@ -686,3 +693,251 @@ vos_obj_evict_by_oid(struct vos_container *cont, daos_unit_oid_t oid)
...
+		return 0;
+
+	D_ASSERT(count > 0);
+	D_ALLOC(vos_hdl, sizeof(*vos_hdl) + sizeof(struct vos_object *) * count);
+	if (vos_hdl == NULL)
+		return -DER_NOMEM;
+
+	vos_bkt_array_init(&bkts);
+	for (i = 0; i < count; i++) {
+		rc = vos_obj_acquire(cont, oids[i], &vos_hdl->vph_objs[i]);
```
Comment (Nasf-Fan) [inline src/vos/vos_obj_cache.c:911]: Why not vos_obj_hold()? that will also pin the object, right?

File: src/vos/vos_obj_index.c [hunk 7]
```diff
@@ -252,9 +252,10 @@ vos_oi_find(struct vos_container *cont, daos_unit_oid_t oid,
 		ilog = &obj->vo_ilog;
 	}
 
-	tmprc = vos_ilog_ts_add(ts_set, ilog, &oid, sizeof(oid));
-
-	D_ASSERT(tmprc == 0); /* Non-zero return for akey only */
+	if (ts_set) {
```
Comment (Nasf-Fan) [inline src/vos/vos_obj_index.c:255]: It seems vos_ilog_ts_add() will handle that internally.

**Review by Nasf-Fan — CHANGES_REQUESTED** [review]

File: src/vos/vos_dtx.c (hunk 2 above)
Comment (NiuYawei) [reply src/vos/vos_dtx.c:2279]: right, if two dae/tx have different buckets, we'll make them commit in two transactions.

File: src/vos/vos_dtx.c (hunk 3 above)
Comment (NiuYawei) [reply src/vos/vos_dtx.c:2344]: I see, thanks!

File: src/vos/vos_dtx.c (hunk 4 above)
Comment (NiuYawei) [reply src/vos/vos_dtx.c:2394]: We need to make it functional first, let's consider optimizing later when necessary.

File: src/vos/vos_obj_cache.c (hunk 6 above)
Comment (NiuYawei) [reply src/vos/vos_obj_cache.c:911]: We don't want to pin object here.

File: src/vos/vos_obj_index.c (hunk 7 above)
Comment (NiuYawei) [reply src/vos/vos_obj_index.c:255]: ok, I'll remove this check.

File: src/vos/vos_io.c (hunk 5 above)
Comment (NiuYawei) [reply src/vos/vos_io.c:2451]: I'm not sure if I follow the 1 item, let's discuss it offline.

File: src/vos/vos_dtx.c (hunk 3 above)
Comment (NiuYawei) [reply src/vos/vos_dtx.c:2344]: hmm, vos_dtx_abort() does need to access object data, but I didn't see how vos_dtx_set_flags() access object data.

File: src/vos/vos_io.c [hunk 8]
```diff
@@ -2429,26 +2434,27 @@ vos_update_end(daos_handle_t ioh, uint32_t pm_ver, daos_key_t *dkey, int err,
 	if (err != 0)
 		goto abort;
 
-	/* The object is already held for md-on-ssd phase2 pool */
-	if (ioc->ic_obj == NULL) {
```
Comment (Nasf-Fan) [inline src/vos/vos_io.c:2433]: why removed the check "if (ioc->ic_obj == NULL)"

File: src/vos/vos_io.c [hunk 9]
```diff
@@ -2617,12 +2623,11 @@ vos_update_begin(daos_handle_t coh, daos_unit_oid_t oid, daos_epoch_t epoch,
...
-		D_ASSERT(rc == 0);
-
-		rc = vos_obj_hold(ioc->ic_cont, ioc->ic_oid, &ioc->ic_epr, ioc->ic_bound,
-				  VOS_OBJ_VISIBLE | VOS_OBJ_CREATE, DAOS_INTENT_UPDATE,
-				  &ioc->ic_obj, ioc->ic_ts_set);
+		/*
+		 * To avoid the complication of adding object ilog to ts_set, we simply pin object
+		 * here without ts_set involved.
+		 */
+		rc = vos_obj_acquire(ioc->ic_cont, ioc->ic_oid, true, &ioc->ic_pinned_obj);
```
Comment (Nasf-Fan) [inline src/vos/vos_io.c:2630]: It seems that the `ioc->ic_pinned_obj` and `ioc->ic_obj` are for the same OID (`ioc->ic_oid`)?

File: src/vos/vos_io.c (hunk 8 above)
Comment (NiuYawei) [reply src/vos/vos_io.c:2433]: I used a separate "ic_pinned_obj" now.

**Pull request merged**